prov/coll/src/coll_coll.c

```diff
@@ -0,0 +1,1278 @@
+	coll_progress_work(util_ep, broadcast_op);
+
+	return FI_SUCCESS;
+err2:
+	free(broadcast_op->data.broadcast.chunk);
+err1:
+	free(broadcast_op);
+	return ret;
+}
+
+ssize_t coll_peer_xfer_complete(struct fid_ep *ep,
+				struct fi_cq_tagged_entry *cqe,
+				fi_addr_t src_addr)
+{
+	struct util_coll_operation *coll_op;
+	struct util_ep *util_ep;
+	struct util_coll_xfer_item *xfer_item;
+
+	xfer_item = cqe->op_context;
+	xfer_item->hdr.state = UTIL_COLL_COMPLETE;
+
+	coll_op = xfer_item->hdr.coll_op;
+	FI_DBG(coll_op->mc->av_set->av->prov, FI_LOG_CQ,
+	       "\tXfer complete: { %p %s Remote: 0x%02x Local: "
+	       "0x%02lx cnt: %d typesize: %ld }\n", xfer_item,
+	       xfer_item->hdr.type == UTIL_COLL_SEND ? "SEND" : "RECV",
+	       xfer_item->remote_rank, coll_op->mc->local_rank,
+	       xfer_item->count, ofi_datatype_size(xfer_item->datatype));
+
+	util_ep = container_of(coll_op->ep, struct util_ep, ep_fid);
+	coll_progress_work(util_ep, coll_op);
+
+	return 0;
+}
+
+ssize_t coll_peer_xfer_error(struct fid_ep *ep, struct fi_cq_err_entry *cqerr)
+{
+	struct util_coll_operation *coll_op;
+	struct util_coll_xfer_item *xfer_item;
+
+	xfer_item = cqerr->op_context;
+	xfer_item->hdr.state = UTIL_COLL_COMPLETE;
+
+	coll_op = xfer_item->hdr.coll_op;
+	FI_DBG(coll_op->mc->av_set->av->prov, FI_LOG_CQ,
+	       "\tXfer error: { %p %s Remote: 0x%02x Local: "
+	       "0x%02lx cnt: %d typesize: %ld }\n", xfer_item,
+	       xfer_item->hdr.type == UTIL_COLL_SEND ? "SEND" : "RECV",
+	       xfer_item->remote_rank, coll_op->mc->local_rank,
+	       xfer_item->count, ofi_datatype_size(xfer_item->datatype));
+
+	// TODO: finish the work with error
+
+	return 0;
+}
+
+int coll_query_collective(struct fid_domain *dom_fid, enum fi_collective_op coll,
+			  struct fi_collective_attr *attr, uint64_t flags)
+{
+	int ret;
+	struct coll_domain *domain = container_of(dom_fid, struct coll_domain,
+						  util_domain.domain_fid);
+	struct fid_domain *peer_domain = domain->peer_domain;
+
+	if (!attr || attr->mode != 0)
+		return -FI_EINVAL;
+
+	switch (coll) {
+	case FI_BARRIER:
+	case FI_ALLGATHER:
+	case FI_SCATTER:
+	case FI_BROADCAST:
+		ret = FI_SUCCESS;
+		break;
+	case FI_ALLREDUCE:
+		if (FI_MIN <= attr->op && FI_BXOR >= attr->op)
+			ret = fi_query_atomic(peer_domain, attr->datatype, attr->op,
+					      &attr->datatype_attr, flags);
+		else
+			return -FI_ENOSYS;
+		break;
+	case FI_ALLTOALL:
+	case FI_REDUCE_SCATTER:
+	case FI_REDUCE:
+	case FI_GATHER:
+	default:
+		return -FI_ENOSYS;
+	}
+
+	if (ret)
+		return ret;
+
+	/* with the currently implemented software based collective operations
+	 * the only restriction is the number of ranks we can address, as
+	 * limited by the size of the rank portion of the collective tag, which
+	 * is 31 bits.  Future collectives may impose further restrictions which
+	 * will need to update the calculation.  For example, operations which
+	 * require dedicated space in the recieve buffer for each rank would
```

Comment:
I guessing most of this code is directly derived from the current code.  Can you update the comments for consistency (mix of /* and // formating, with sometimes a blank line to start the comment), and also fix the spelling of receive here?

Reply:
Will do.